// Code generated by protoc-gen-validate. DO NOT EDIT.
// source: saacs/example/v0/nested.proto

package v0

import (
	"bytes"
	"errors"
	"fmt"
	"net"
	"net/mail"
	"net/url"
	"regexp"
	"sort"
	"strings"
	"time"
	"unicode/utf8"

	"google.golang.org/protobuf/types/known/anypb"
)

// ensure the imports are used
var (
	_ = bytes.MinRead
	_ = errors.New("")
	_ = fmt.Print
	_ = utf8.UTFMax
	_ = (*regexp.Regexp)(nil)
	_ = (*strings.Reader)(nil)
	_ = net.IPv4len
	_ = time.Duration(0)
	_ = (*url.URL)(nil)
	_ = (*mail.Address)(nil)
	_ = anypb.Any{}
	_ = sort.Sort
)

// Validate checks the field values on ItemWithNestedKey with the rules defined
// in the proto definition for this message. If any rules are violated, the
// first error encountered is returned, or nil if there are no violations.
func (m *ItemWithNestedKey) Validate() error {
	return m.validate(false)
}

// ValidateAll checks the field values on ItemWithNestedKey with the rules
// defined in the proto definition for this message. If any rules are
// violated, the result is a list of violation errors wrapped in
// ItemWithNestedKeyMultiError, or nil if none found.
func (m *ItemWithNestedKey) ValidateAll() error {
	return m.validate(true)
}

func (m *ItemWithNestedKey) validate(all bool) error {
	if m == nil {
		return nil
	}

	var errors []error

	// no validation rules for CollectionId

	// no validation rules for Id

	if all {
		switch v := interface{}(m.GetNested()).(type) {
		case interface{ ValidateAll() error }:
			if err := v.ValidateAll(); err != nil {
				errors = append(errors, ItemWithNestedKeyValidationError{
					field:  "Nested",
					reason: "embedded message failed validation",
					cause:  err,
				})
			}
		case interface{ Validate() error }:
			if err := v.Validate(); err != nil {
				errors = append(errors, ItemWithNestedKeyValidationError{
					field:  "Nested",
					reason: "embedded message failed validation",
					cause:  err,
				})
			}
		}
	} else if v, ok := interface{}(m.GetNested()).(interface{ Validate() error }); ok {
		if err := v.Validate(); err != nil {
			return ItemWithNestedKeyValidationError{
				field:  "Nested",
				reason: "embedded message failed validation",
				cause:  err,
			}
		}
	}

	// no validation rules for Value

	if len(errors) > 0 {
		return ItemWithNestedKeyMultiError(errors)
	}

	return nil
}

// ItemWithNestedKeyMultiError is an error wrapping multiple validation errors
// returned by ItemWithNestedKey.ValidateAll() if the designated constraints
// aren't met.
type ItemWithNestedKeyMultiError []error

// Error returns a concatenation of all the error messages it wraps.
func (m ItemWithNestedKeyMultiError) Error() string {
	var msgs []string
	for _, err := range m {
		msgs = append(msgs, err.Error())
	}
	return strings.Join(msgs, "; ")
}

// AllErrors returns a list of validation violation errors.
func (m ItemWithNestedKeyMultiError) AllErrors() []error { return m }

// ItemWithNestedKeyValidationError is the validation error returned by
// ItemWithNestedKey.Validate if the designated constraints aren't met.
type ItemWithNestedKeyValidationError struct {
	field  string
	reason string
	cause  error
	key    bool
}

// Field function returns field value.
func (e ItemWithNestedKeyValidationError) Field() string { return e.field }

// Reason function returns reason value.
func (e ItemWithNestedKeyValidationError) Reason() string { return e.reason }

// Cause function returns cause value.
func (e ItemWithNestedKeyValidationError) Cause() error { return e.cause }

// Key function returns key value.
func (e ItemWithNestedKeyValidationError) Key() bool { return e.key }

// ErrorName returns error name.
func (e ItemWithNestedKeyValidationError) ErrorName() string {
	return "ItemWithNestedKeyValidationError"
}

// Error satisfies the builtin error interface
func (e ItemWithNestedKeyValidationError) Error() string {
	cause := ""
	if e.cause != nil {
		cause = fmt.Sprintf(" | caused by: %v", e.cause)
	}

	key := ""
	if e.key {
		key = "key for "
	}

	return fmt.Sprintf(
		"invalid %sItemWithNestedKey.%s: %s%s",
		key,
		e.field,
		e.reason,
		cause)
}

var _ error = ItemWithNestedKeyValidationError{}

var _ interface {
	Field() string
	Reason() string
	Key() bool
	Cause() error
	ErrorName() string
} = ItemWithNestedKeyValidationError{}

// Validate checks the field values on Nested with the rules defined in the
// proto definition for this message. If any rules are violated, the first
// error encountered is returned, or nil if there are no violations.
func (m *Nested) Validate() error {
	return m.validate(false)
}

// ValidateAll checks the field values on Nested with the rules defined in the
// proto definition for this message. If any rules are violated, the result is
// a list of violation errors wrapped in NestedMultiError, or nil if none found.
func (m *Nested) ValidateAll() error {
	return m.validate(true)
}

func (m *Nested) validate(all bool) error {
	if m == nil {
		return nil
	}

	var errors []error

	// no validation rules for Part1

	// no validation rules for Part2

	if len(errors) > 0 {
		return NestedMultiError(errors)
	}

	return nil
}

// NestedMultiError is an error wrapping multiple validation errors returned by
// Nested.ValidateAll() if the designated constraints aren't met.
type NestedMultiError []error

// Error returns a concatenation of all the error messages it wraps.
func (m NestedMultiError) Error() string {
	var msgs []string
	for _, err := range m {
		msgs = append(msgs, err.Error())
	}
	return strings.Join(msgs, "; ")
}

// AllErrors returns a list of validation violation errors.
func (m NestedMultiError) AllErrors() []error { return m }

// NestedValidationError is the validation error returned by Nested.Validate if
// the designated constraints aren't met.
type NestedValidationError struct {
	field  string
	reason string
	cause  error
	key    bool
}

// Field function returns field value.
func (e NestedValidationError) Field() string { return e.field }

// Reason function returns reason value.
func (e NestedValidationError) Reason() string { return e.reason }

// Cause function returns cause value.
func (e NestedValidationError) Cause() error { return e.cause }

// Key function returns key value.
func (e NestedValidationError) Key() bool { return e.key }

// ErrorName returns error name.
func (e NestedValidationError) ErrorName() string { return "NestedValidationError" }

// Error satisfies the builtin error interface
func (e NestedValidationError) Error() string {
	cause := ""
	if e.cause != nil {
		cause = fmt.Sprintf(" | caused by: %v", e.cause)
	}

	key := ""
	if e.key {
		key = "key for "
	}

	return fmt.Sprintf(
		"invalid %sNested.%s: %s%s",
		key,
		e.field,
		e.reason,
		cause)
}

var _ error = NestedValidationError{}

var _ interface {
	Field() string
	Reason() string
	Key() bool
	Cause() error
	ErrorName() string
} = NestedValidationError{}
